internal/controller/reconciler_test.go

```diff
@@ -0,0 +1,229 @@
+
+			return nil
+		}
+	}
+
+	getReturnsNotFoundErrorForHR := func(hr *v1beta1.HTTPRoute) getFunc {
+		return func(
+			ctx context.Context,
+			nsname types.NamespacedName,
+			object client.Object,
+			option ...client.GetOption,
+		) error {
+			Expect(object).To(BeAssignableToTypeOf(&v1beta1.HTTPRoute{}))
+			Expect(nsname).To(Equal(client.ObjectKeyFromObject(hr)))
+
+			return apierrors.NewNotFound(schema.GroupResource{}, "not found")
+		}
+	}
+
+	BeforeEach(func() {
+		fakeGetter = &controllerfakes.FakeGetter{}
+		eventCh = make(chan interface{})
+	})
+
+	Describe("Normal cases", func() {
+		reconcile := func(nsname types.NamespacedName) {
+			result, err := reconciler.Reconcile(context.Background(), reconcile.Request{NamespacedName: nsname})
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(result).To(Equal(reconcile.Result{}))
+		}
+
+		When("Reconciler doesn't have a filter", func() {
+			BeforeEach(func() {
+				reconciler = controller.NewReconcilerImpl(
+					fakeGetter, controller.Options{
+						ObjectType: &v1beta1.HTTPRoute{},
+						EventCh:    eventCh,
+					},
+					"HTTPRoute")
+			})
+
+			It("should upsert HTTPRoute", func() {
+				fakeGetter.GetCalls(getReturnsHRForHR(hr1))
+
+				go reconcile(hr1NsName)
+
+				Eventually(eventCh).Should(Receive(Equal(&events.UpsertEvent{Resource: hr1})))
+			})
+
+			It("should delete HTTPRoute", func() {
+				fakeGetter.GetCalls(getReturnsNotFoundErrorForHR(hr1))
+
+				go reconcile(hr1NsName)
+
+				Eventually(eventCh).Should(Receive(Equal(&events.DeleteEvent{
+					NamespacedName: hr1NsName,
+					Type:           &v1beta1.HTTPRoute{},
+				})))
+			})
+		})
+
+		When("Reconciler has a NamespacedNameFilter", func() {
+			BeforeEach(func() {
+				filter := func(nsname types.NamespacedName) (bool, string) {
+					if nsname != hr1NsName {
+						return false, "ignore"
+					}
+					return true, ""
+				}
+
+				reconciler = controller.NewReconcilerImpl(
+					fakeGetter,
+					controller.Options{
+						ObjectType:           &v1beta1.HTTPRoute{},
+						EventCh:              eventCh,
+						NamespacedNameFilter: filter,
+					},
+					"HTTPRoute",
+				)
+			})
+
+			When("HTTPRoute is not ignored", func() {
+				It("should upsert HTTPRoute", func() {
+					fakeGetter.GetCalls(getReturnsHRForHR(hr1))
+
+					go reconcile(hr1NsName)
+
+					Eventually(eventCh).Should(Receive(Equal(&events.UpsertEvent{Resource: hr1})))
+				})
+
+				It("should delete HTTPRoute", func() {
+					fakeGetter.GetCalls(getReturnsNotFoundErrorForHR(hr1))
+
+					go reconcile(hr1NsName)
+
+					Eventually(eventCh).Should(Receive(Equal(&events.DeleteEvent{
+						NamespacedName: hr1NsName,
+						Type:           &v1beta1.HTTPRoute{},
+					})))
+				})
+			})
+
+			When("HTTPRoute is ignored", func() {
+				It("should not upsert HTTPRoute", func() {
+					fakeGetter.GetCalls(getReturnsHRForHR(hr2))
+
+					reconcile(hr2NsName)
+					// if reconcile doesn't block, it means that the event was not sent to the eventCh,
```

Comment:
Can we test this more explicitly by using the following matcher?

```
Consistently(eventCh).ShouldNot(Receive())
```

Reply:
I added Consistently.

Additionally, to ensure that reconciling always finishes (never hangs), I made `reconcile` return a channel that publishes the result of reconciliation. If we receive on that channel, that means reconciliation actually happened.